### PR TITLE
Implementation of CpuUtilizationBased Recommender

### PR DESCRIPTION
**Pull request by neerajb**
- Impl of cpuutilbasedrecommender. 
- It can simulate the HPA algorithm for the datapoints that are fetched by the scraper. 
- It runs a binary search on the min, max target values to find the optimal value without breaching the redLine utilization.
- Introduction of a QueryRangeSplitter in scraper to split the query by a time.Duration to avoid loading too many samples in P8s memory.

### CODE REVIEW

File: pkg/metrics/scraper.go [hunk 1]
```diff
@@ -10,32 +10,39 @@ import (
 	"github.com/prometheus/common/model"
 )
 
+type DataPoint struct {
```
Comment (bharathguvvala) [inline pkg/metrics/scraper.go:13]: SamplePair can be used here? https://github.com/prometheus/common/blob/e27922a100211b71acc8e65327b61d5fe22e2a77/model/value_float.go#L69

Comment (neerajb) [reply pkg/metrics/scraper.go:13]: Didn't want to expose p8s client data structures in the scraper interface. As this is a generic interface that can be extended for non promql datasources in future.

File: pkg/reco/reco.go [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+		return []metrics.DataPoint{}, 0, 0, nil
+	}
+	if targetUtilization < 1 || targetUtilization > 100 {
+		return []metrics.DataPoint{}, 0, 0, errors.New(fmt.Sprintf("Invalid value of target utilization: %v."+
+			" Value should be between 1 and 100", targetUtilization))
+	}
+
+	simulatedDataPoints := make([]metrics.DataPoint, len(dataPoints))
+
+	currentReplicas := math.Ceil((dataPoints[0].Value * 100) / float64(targetUtilization) / perPodResources)
```
Comment (bharathguvvala) [inline pkg/reco/reco.go:115]: we are assuming here the perPodResources has been constant and is same in the past. Changes to perPodResources in the past is going to mess this up.

Comment (neerajb) [reply pkg/reco/reco.go:115]: This is intentional. We are only getting the total cpu utilisation demand curve from the past. Per pod resources don't matter. We only need to get the current perPodResources.

File: pkg/reco/reco.go [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+	return &v1alpha1.HPAConfiguration{Min: minReplicas, Max: maxReplicas, TargetMetricValue: optimalTargetUtil}, nil
+}
+
+type TimerEvent struct {
+	Timestamp time.Time
+	Delta     float64
+}
+
+func (c *CpuUtilizationBasedRecommender) simulateHPA(dataPoints []metrics.DataPoint,
```
Comment (bharathguvvala) [inline pkg/reco/reco.go:100]: Please add some comments to this method to elaborate on the approach/steps.

Comment (neerajb) [reply pkg/reco/reco.go:107]: Done.

**Review by bharathguvvala — APPROVED** [review]

**Pull request merged**